### PR TITLE
Feat: As a customer, I want to get the stock notification for a product when it's out of stock, and I can manage my subscription

### LINKED ISSUE #2285: Notification preference
As a customer, I want enable/disable the notification, so that I can receive the notification via my preferred channel(s)

- Create a new page called "Notification Channels" under "My account"
- Show the list of all the available channels and show a switch for each channel
- Be able to enable/disable a channel by clicking the switch
-- After click a switch of a channel, disable all the checkbox until getting the response from server, then update the page accordingly
- Anonymous user will be redirected to register/login page when he/she try to enter the "Notification Channels"
- By default, don't show "Site_Message" channel
- Please refer to the mock-up about the UI design

*NFR*
- This feature should be configurable

### PR DESCRIPTION
**Pull request by Jerify02**
Introduce components, selectors, ngrx and occ-services for notification preference, my interest and stock notification.

Breaking Changes:
Add new states in `user-state.ts`.

Closed #2285, #2268 and #5123

### CODE REVIEW

File: projects/storefrontlib/src/cms-components/myaccount/my-interests/my-interests.component.html [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+                    </div>
+                  </div>
+                </td>
+                <td>
+                  <div class="cx-actions cx-product-interests-remove-button">
+                    <button
+                      type="button"
+                      class="link cx-product-interests-remove-btn"
+                      [class.disabled]="
+                        (getInterestsloading$ | async) ||
```
Comment (dunqan) [inline projects/storefrontlib/src/cms-components/myaccount/my-interests/my-interests.component.html]: Do we need `getInterestsloading$` here, as it seems it's taken into account above?

File: projects/storefrontlib/src/cms-components/myaccount/my-interests/my-interests.component.html [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+                        (getRemoveInterestsloading$ | async)
+                      "
+                      [disabled]="
+                        (getInterestsloading$ | async) ||
```
Comment (dunqan) [inline projects/storefrontlib/src/cms-components/myaccount/my-interests/my-interests.component.html]: Do we need `getInterestsloading$` here, as it seems it's taken into account above?

File: projects/storefrontlib/src/cms-components/myaccount/notification-preference/notification-preference.component.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import { map, tap } from 'rxjs/operators';
+
+@Component({
+  selector: 'cx-notification-preference',
+  templateUrl: './notification-preference.component.html',
+})
+export class NotificationPreferenceComponent implements OnInit {
+  preferences$: Observable<NotificationPreference[]>;
+  isLoading$: Observable<boolean>;
+  preferences: NotificationPreference[] = [];
```
Comment (dunqan) [inline projects/storefrontlib/src/cms-components/myaccount/notification-preference/notification-preference.component.ts]: What about marking this `protected`?

File: projects/storefrontlib/src/cms-components/myaccount/notification-preference/notification-preference.component.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import {
+  NotificationPreference,
+  UserNotificationPreferenceService,
+} from '@spartacus/core';
+import { Observable, combineLatest } from 'rxjs';
+import { map, tap } from 'rxjs/operators';
+
+@Component({
+  selector: 'cx-notification-preference',
+  templateUrl: './notification-preference.component.html',
```
Comment (dunqan) [inline projects/storefrontlib/src/cms-components/myaccount/notification-preference/notification-preference.component.ts]: Can we add `  changeDetection: ChangeDetectionStrategy.OnPush,` ?

File: projects/storefrontlib/src/cms-components/myaccount/my-interests/my-interests.component.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  ProductInterestSearchResult,
+  ProductInterestEntryRelation,
+  PaginationModel,
+  TranslationService,
+} from '@spartacus/core';
+import { tap, map } from 'rxjs/operators';
+import { ICON_TYPE } from '../../misc/icon/icon.model';
+@Component({
+  selector: 'cx-my-interests',
+  templateUrl: './my-interests.component.html',
```
Comment (dunqan) [inline projects/storefrontlib/src/cms-components/myaccount/my-interests/my-interests.component.ts]: Can we use  `changeDetection: ChangeDetectionStrategy.OnPush,`?

File: projects/storefrontlib/src/cms-components/product/stock-notification/stock-notification-dialog/stock-notification-dialog.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    expect(modalService.dismissActiveModal).toHaveBeenCalled();
+  });
+
+  it('should be able to close dialog by OK button', () => {
+    fixture.detectChanges();
+    el.query(By.css('.btn-ok')).nativeElement.click();
+    expect(modalService.dismissActiveModal).toHaveBeenCalled();
+  });
+
+  it('should able to reset the state in destory', () => {
```
Comment (dunqan) [inline projects/storefrontlib/src/cms-components/product/stock-notification/stock-notification-dialog/stock-notification-dialog.component.spec.ts]: Typo

File: projects/storefrontlib/src/cms-components/product/stock-notification/stock-notification.component.ts [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+  selector: 'cx-stock-notification',
+  templateUrl: './stock-notification.component.html',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class StockNotificationComponent implements OnInit, OnDestroy {
+  subscribed$: Observable<boolean>;
+  prefsEnabled$: Observable<boolean>;
+  outOfStock$: Observable<boolean>;
+  subscribeSuccess$: Observable<boolean>;
+  unsubscribeLoading$: Observable<boolean>;
```
Comment (dunqan) [inline projects/storefrontlib/src/cms-components/product/stock-notification/stock-notification.component.ts]: This name is not very obvious, what about `isRemoveProdutInterestLoading$` or something like that?

File: projects/storefrontlib/src/cms-components/product/stock-notification/stock-notification.component.ts [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+import { ModalService } from '../../../shared/components/modal/modal.service';
+import { StockNotificationDialogComponent } from './stock-notification-dialog/stock-notification-dialog.component';
+
+@Component({
+  selector: 'cx-stock-notification',
+  templateUrl: './stock-notification.component.html',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class StockNotificationComponent implements OnInit, OnDestroy {
+  subscribed$: Observable<boolean>;
```
Comment (dunqan) [inline projects/storefrontlib/src/cms-components/product/stock-notification/stock-notification.component.ts]: `hasProductInterests$`?

File: projects/storefrontlib/src/cms-components/myaccount/my-interests/my-interests.component.ts [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+  pageChange(page: number): void {
+    this.productInterestService.loadProductInterests(
+      this.DEFAULT_PAGE_SIZE,
+      page,
+      this.sortMapping[this.sort]
+    );
+  }
+
+  ngOnDestroy(): void {
+    this.productInterestService.clearProductInterests();
```
Comment (dunqan) [inline projects/storefrontlib/src/cms-components/myaccount/my-interests/my-interests.component.ts]: we are doing it twice, in `onInit` and here. Is it necessary?

File: projects/core/src/occ/adapters/user/occ-user-interests.adapter.ts [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    userId: string,
+    pageSize?: number,
+    currentPage?: number,
+    sort?: string,
+    productCode?: string,
+    notificationType?: NotificationType
+  ): Observable<ProductInterestSearchResult> {
+    let params = new HttpParams().set('sort', sort ? sort : 'name:asc');
+    if (pageSize) {
+      params = params.set('pageSize', pageSize.toString());
```
[the remark this answers is not in the document]
Comment (Jerify02) [reply projects/core/src/occ/adapters/user/occ-user-interests.adapter.ts]: Yes, the number of params seems a little more, and all these params are optional, but I don't know what you mean? For example?

File: projects/storefrontlib/src/cms-components/myaccount/my-interests/my-interests.component.html (hunk 1 above)
Comment (Jerify02) [reply projects/storefrontlib/src/cms-components/myaccount/my-interests/my-interests.component.html]: removed.

File: projects/storefrontlib/src/cms-components/myaccount/my-interests/my-interests.component.html (hunk 2 above)
Comment (Jerify02) [reply projects/storefrontlib/src/cms-components/myaccount/my-interests/my-interests.component.html]: I think it is necessary, it is a deletion operation after all.

File: projects/storefrontlib/src/cms-components/product/stock-notification/stock-notification.component.ts (hunk 8 above)
Comment (Jerify02) [reply projects/storefrontlib/src/cms-components/product/stock-notification/stock-notification.component.ts]: I think it is OK, `subscribed` can indicate whether current product is subscribed as a product interest.

File: projects/storefrontlib/src/cms-components/myaccount/my-interests/my-interests.component.ts (hunk 9 above)
Comment (Jerify02) [reply projects/storefrontlib/src/cms-components/myaccount/my-interests/my-interests.component.ts]: removed.

File: projects/core/src/occ/adapters/user/user-occ.module.ts [hunk 11]
```diff
@@ -35,6 +41,16 @@ import { OccUserAdapter } from './occ-user.adapter';
...
     },
     { provide: UserOrderAdapter, useClass: OccUserOrderAdapter },
+    {
+      provide: UserNotificationPreferenceAdapter,
+      useClass: OccUserNotificationPreferenceAdapter,
+    },
+    { provide: UserInterestsAdapter, useClass: OccUserInterestsAdapter },
+    {
+      provide: PRODUCT_INTERESTS_NORMALIZER,
+      useClass: OccUserInterestsNormalizer,
```
[the remark this answers is not in the document]
Comment (Jerify02) [reply projects/core/src/occ/adapters/user/user-occ.module.ts]: fixed

**Comment by dunqan** [comment]
In general, we are trying to:

  - avoid the duplicate state in the store
  - load only what's needed for performance reasons (because spartacus, in contrast to the old 
accelerator, works on a client-side, so each data that is not needed is making the experience worse) 

I would suggest trying to load (and put to the store) only product codes for product interests entries (currently we fetch and keep product data in the store) and get the product using product facade/service?

The products are quite a big chunk of our store, and it's the first one that we want to optimize and keep normalized (only once in the store).

Also, we try to avoid using `fields=FULL` and prefer to explicitly specify what's needed, because:
  - FULL from the frontend perspective is unknown, which makes some additional performance improvements impossible, 
  -  we basically have two sources of truth for fields definition
So if it is possible to substitute FULL with some explicit fields, would also be good.

File: projects/storefrontstyles/scss/components/myaccount/_my-interests.scss [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+        button.link {
+          font-weight: var(--cx-font-weight-bold);
+          font-size: var(--cx-font-size, 0.875rem);
+        }
+      }
+
+      .cx-code {
+        font-size: var(--cx-font-size, 0.875rem);
+        font-weight: var(--cx-font-weight-normal);
+        line-height: var(--cx-line-height, 1.2222222222);
```
Comment (developpeurweb) [inline projects/storefrontstyles/scss/components/myaccount/_my-interests.scss]: we have discontinued the use of css variables for everything except colors, the line-height value can be assigned directly, also for fonts we're using the `type()` mixin, as per table below. Therefore font-size and font-weight can be replaced in all instance by the proper mixin.

Screen Shot 2019-09-27 at 10 59 00 AM

File: projects/storefrontstyles/scss/components/myaccount/_my-interests.scss [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+
+      @include media-breakpoint-up(md) {
+        &:last-child {
+          text-align: center;
+        }
+      }
+
+      .cx-name {
+        a {
+          color: $dark;
```
Comment (developpeurweb) [inline projects/storefrontstyles/scss/components/myaccount/_my-interests.scss]: The use of SASS variables for colors has been discontinued. All color values should be assigned as **CSS variables** therefore: `$dark` should be `--cx-color-dark`

Screen Shot 2019-11-21 at 11 56 11 AM

Comment (Jerify02) [reply projects/storefrontstyles/scss/components/myaccount/_my-interests.scss]: Updated.

File: projects/storefrontstyles/scss/components/myaccount/_my-interests.scss (hunk 12 above)
Comment (maggielin0) [reply projects/storefrontstyles/scss/components/myaccount/_my-interests.scss]: update.

**Pull request merged**